### PR TITLE
[CS] Add contextual Bool type to switch-case `where` clause 

### LINKED ISSUE #63746: Bad type conversion for switch-case `where` clause crashes compiler
This crashes the compiler:

```swift
let fn = {
  switch 0 {
    case 1 where 0:
      ()
  }
}
```

### PR DESCRIPTION
**Pull request by hamishknight**
Seems this was accidentally omitted, which would crash in CSApply for any non-Bool-convertible type.

Resolves #63746

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by xedin — APPROVED** [review]

File: test/Constraints/diagnostics.swift [hunk 1]
```diff
@@ -1531,6 +1531,16 @@ func testUnwrapFixIts(x: Int?) throws {
   let _: Int = try! .optionalThrowsMember ?? 0
 }
 
+// https://github.com/apple/swift/issues/63746
+func issue63746() {
+  let fn = {
+    switch 0 {
+      case 1 where 0: // expected-error {{type 'Int' cannot be used as a boolean; test for '!= 0' instead}}
```
Comment (AnthonyLatsis) [inline test/Constraints/diagnostics.swift:1538]: As an aside, this fix-it is useless for literals.

Comment (hamishknight) [reply test/Constraints/diagnostics.swift:1538]: Heh yeah, filed https://github.com/apple/swift/issues/63753

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**